src/main/java/no/digipost/monitoring/servlet/ActiveRequestsFilter.java

```diff
@@ -0,0 +1,145 @@
+import java.util.function.Predicate;
+
+import static java.util.Collections.emptyList;
+import static java.util.stream.Collectors.toList;
+
+public class ActiveRequestsFilter implements Filter {
+    final Map<UUID, RequestMetaInfo> activeRequests = new ConcurrentHashMap<>();
+    private final Config config;
+
+    public ActiveRequestsFilter(MeterRegistry registry, ActiveRequestsFilter.Config config) {
+        this.config = config;
+        registry.gauge("app_http_requests_active", activeRequests, Map::size);
+        registry.gauge("app_http_requests_max", config.getMaxThreads());
+
+        registry.gauge("app_http_requests_longrunning", emptyList(), activeRequests,
+                (Map<UUID, RequestMetaInfo> activeRequests) -> getLongRunning(activeRequests, config.longRunningThreshold).size());
+    }
+
+    @Override
+    public void doFilter(final ServletRequest servletRequest, final ServletResponse servletResponse, final FilterChain filterChain) throws IOException, ServletException {
+        UUID requestId = UUID.randomUUID();
+        RequestMetaInfo metaInfo = createRequestMetaInfo(servletRequest, Instant.now());
+        try {
+            activeRequests.put(requestId, metaInfo);
+            filterChain.doFilter(servletRequest, servletResponse);
+        } finally {
+            activeRequests.remove(requestId);
+        }
+    }
+
+    @Override
+    public void init(FilterConfig filterConfig) {}
+
+    @Override
+    public void destroy() {}
+
+    private boolean excludedRequest(RequestMetaInfo requestMetaInfo) {
+        return config.longRunningExclusions.stream().anyMatch(p -> p.test(requestMetaInfo));
+    }
+
+    private List<RequestMetaInfo> getLongRunning(Map<UUID, RequestMetaInfo> activeRequests, Duration longRunningThreshold) {
+        Instant deadline = Instant.now().minus(longRunningThreshold);
+        return activeRequests.values().stream()
+                .filter(m -> m.isOlderThan(deadline) && !excludedRequest(m))
+                .collect(toList());
+    }
+
+    private static RequestMetaInfo createRequestMetaInfo(ServletRequest servletRequest, Instant now) {
+        if (servletRequest instanceof HttpServletRequest) {
+            HttpServletRequest httpReq = (HttpServletRequest) servletRequest;
+            String path = httpReq.getRequestURI();
+            String method = httpReq.getMethod();
+            return new RequestMetaInfo(path, method, now);
+        } else {
+            return new RequestMetaInfo();
+        }
+    }
+
+
+    public static class RequestMetaInfo {
+        public final String path;
+        public final String method;
+        public final Instant requestTime;
+
+        public RequestMetaInfo(String path, String method, Instant requestTime) {
+            this.path = path;
+            this.method = method;
+            this.requestTime = requestTime;
+        }
+
+        public RequestMetaInfo() {
+            this("?","?", Instant.now());
```

Comment:
Should we have a clock?

Reply:
Maybe it doesn’t matter since you are only interested in the relative time.

Reply:
I had one at first for the unit test, but found I didn't really need it so I removed it 🤦 , can re-introduce if you want?